File: Python/Flatten Binary Tree to Linked List.py
# Given a binary tree, flatten it to a linked list in-place.

# For example,
# Given

         # 1
        # / \
       # 2   5
      # / \   \
     # 3   4   6
# The flattened tree should look like:
   # 1
    # \
     # 2
      # \
       # 3
        # \
         # 4
          # \
           # 5
            # \
             # 6
			 
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return nothing, do it in place
	def flatten(self, root):
		if root == None:
			return
		self.convert(root)
	def convert(self,root):
		if root.left == None and root.right == None:
			return root
		rHead = None
		if root.right != None:
			rHead = self.convert(root.right)
		p = root
		if root.left != None:
			lHead = self.convert(root.left)
			root.right = lHead
			lHead.left = None
			root.left = None
			while p.right != None:
				p = p.right
		if rHead != None:
			p.right = rHead
			rHead.left = None
		return root

File: Java/Flatten Binary Tree to Linked List.java
class Flatten Binary Tree to Linked List{
	public void flatten(TreeNode root) {
        iterative(root);
    }
    private TreeNode recursive(TreeNode root){
    	if(root == null)
    		return null;
    	TreeNode lHead = recursive(root.left);
    	TreeNode rHead = recursive(root.right);
    	root.left = null;//this is important
    	if(lHead != null){
    		root.right = lHead;
    		while(lHead.right != null)
    			lHead = lHead.right;
    		lHead.right = rHead;
    	}
    	else
    		root.right = rHead;
    	return root;
    }
    private void iterative(TreeNode root){
    	if(root == null)
    		return;
    	LinkedList<TreeNode> stack = new LinkedList<TreeNode>();
    	stack.push(root);
    	TreeNode pre = null;
    	while(!stack.isEmpty()){
    		TreeNode current = stack.pop();
            if(current.right != null)
                stack.push(current.right);
            if(current.left != null)
                stack.push(current.left);
            if(pre != null){
                pre.left = null;
                pre.right = current;
            }
            pre = current;
    	}
    }
}